sdk/include/reactos/drivers/videoprt/vpcfgcnv.h

```diff
@@ -0,0 +1,864 @@
+    if ( (ResourceList->Version <  1) ||
+        ((ResourceList->Version == 1) && (ResourceList->Revision <= 1)) )
+    {
+        TEST_DPRINT("IsNewConfigData:FALSE - Version %lu, Revision %lu\n",
+                ResourceList->Version, ResourceList->Revision);
+        return FALSE;
+    }
+
+    /* This should be a new configuration data */
+    TEST_DPRINT("IsNewConfigData:TRUE\n");
+    return TRUE;
+}
+
+FORCEINLINE
+BOOLEAN
+IsNewConfigDataFull(
+    _In_ PVOID ConfigurationData,
+    _In_ ULONG ConfigurationDataLength)
+{
+    if (!ConfigurationData)
+    {
+        TEST_DPRINT("IsNewConfigDataFull:FALSE - ConfigurationData == NULL\n");
+        return FALSE;
+    }
+
+    if (ConfigurationDataLength < FIELD_OFFSET(CM_FULL_RESOURCE_DESCRIPTOR,
+                                               PartialResourceList.PartialDescriptors))
+    {
+        TEST_DPRINT("IsNewConfigDataFull:FALSE - ConfigurationDataLength == %lu\n",
+                ConfigurationDataLength);
+        return FALSE;
+    }
+
+    /* Verify the actual new configuration data */
+    return IsNewConfigData(
+                GET_LEGACY_DATA(ConfigurationData),
+                GET_LEGACY_DATA_LEN(ConfigurationDataLength));
+}
+
+
+/**
+ * @brief
+ * Given a CM_PARTIAL_RESOURCE_LIST, obtain pointers to resource descriptors
+ * for legacy video configuration: interrupt, control and cursor I/O ports,
+ * and framebuffer memory descriptors.  In addition, retrieve any
+ * device-specific resource present.
+ **/
+FORCEINLINE
+VOID
+GetVideoData(
+    _In_ PCM_PARTIAL_RESOURCE_LIST ResourceList,
+    _Out_opt_ PCM_PARTIAL_RESOURCE_DESCRIPTOR* Interrupt,
+    _Out_opt_ PCM_PARTIAL_RESOURCE_DESCRIPTOR* ControlPort,
+    _Out_opt_ PCM_PARTIAL_RESOURCE_DESCRIPTOR* CursorPort,
+    _Out_ PCM_PARTIAL_RESOURCE_DESCRIPTOR* FrameBuffer,
+    _Out_opt_ PCM_PARTIAL_RESOURCE_DESCRIPTOR* DeviceSpecific)
+{
+    PCM_PARTIAL_RESOURCE_DESCRIPTOR Descriptor;
+    ULONG PortCount = 0, IntCount = 0, MemCount = 0;
+    ULONG i;
+
+    /* Initialize the return values */
+    if (Interrupt)   *Interrupt   = NULL;
+    if (ControlPort) *ControlPort = NULL;
+    if (CursorPort)  *CursorPort  = NULL;
+    *FrameBuffer = NULL;
+    if (DeviceSpecific) *DeviceSpecific = NULL;
+
+    TEST_DPRINT("GetVideoData\n");
+
+    for (i = 0; i < ResourceList->Count; ++i)
+    {
+        Descriptor = &ResourceList->PartialDescriptors[i];
+
+        switch (Descriptor->Type)
+        {
+            case CmResourceTypePort:
+            {
+                TEST_DPRINT("    CmResourceTypePort\n");
+                /* We only check for memory I/O ports */
+                // if (!(Descriptor->Flags & CM_RESOURCE_PORT_MEMORY))
+                if (Descriptor->Flags & CM_RESOURCE_PORT_IO)
+                    break;
+
+                /* If more than two memory I/O ports
+                 * have been encountered, ignore them */
+                if (PortCount > 2)
```

Comment:
Wait, why?

Reply:
This `GetVideoData()` function is a helper to convert a boot framebuffer video resource descriptor from what I call a _new-format_ descriptor to the (half-documented) legacy video configuration, see the `ConvertVideoDataToLegacyConfigData()` function below.

[The legacy video configuration](https://github.com/reactos/reactos/pull/6659/files#diff-85613155ac956677483f1b5b35a54925ce570fd7e1cbbaea0254eb08c4d192e4R79) lists data for one interrupt, two ports (control/cursor), and one memory "framebuffer" area associated to the video resource.
The new resource descriptor format for describing the boot framebuffer resources, is based on the standard `CM_PARTIAL_RESOURCE_LIST` with one or more `CM_PARTIAL_RESOURCE_DESCRIPTOR`'s.
The idea is that in the list of resource descriptors, the 1st interrupt resource would correspond to the legacy interrupt data, the _**first two memory IO port**_ resources would correspond to the two legacy ports in a given order (control, then cursor), and the 1st memory resource would correspond to the legacy framebuffer area.
This is a convention and assumption I make, since I'm basically defining the usage of the new-format descriptor for this usage of interest.

(NOTE: I mentioned the _**first two memory IO port**_ resources (amongst all the possible IO ports that could be enumerated), because this control and cursor IO ports are usually memory-mapped ports. This is the case in MIPS architectures; this may be seen as more questionable on e.g. x86...)

The new descriptor for the boot framebuffer resources is what is intended to be used, in order to be also able to specify a custom [`CM_FRAMEBUF_DEVICE_DATA` structure](https://github.com/reactos/reactos/pull/6658/files#diff-a4e51a88bd7c7855d4af1dc22b1a599110b8b92bc27b824ff30b54bec2e66446R16) (see PR #6658) for specifying the format of the boot framebuffer device.
(See also what I mention at [line 27](https://github.com/reactos/reactos/pull/6659/files#diff-85613155ac956677483f1b5b35a54925ce570fd7e1cbbaea0254eb08c4d192e4R27) and [line 56](https://github.com/reactos/reactos/pull/6659/files#diff-85613155ac956677483f1b5b35a54925ce570fd7e1cbbaea0254eb08c4d192e4R56) above in the file.)